### PR TITLE
[#480][pulsar-flink] Rewrite extractTopicName()

### PR DESCRIPTION
**Pull request by imaffe**
TEST Result @nlu90 :

### CODE REVIEW

File: pulsar-flink-connector/src/main/java/org/apache/flink/streaming/connectors/pulsar/catalog/util/PulsarCatalogSupport.java [hunk 1]
```diff
@@ -49,7 +49,7 @@
 /** catalog support. */
 public class PulsarCatalogSupport {
 
-    private static final String TABLE_PREFIX = "table_";
+    public static final String TABLE_PREFIX = "table_";
```
Comment (imaffe) [inline pulsar-flink-connector/src/main/java/org/apache/flink/streaming/connectors/pulsar/catalog/util/PulsarCatalogSupport.java:52]: This change is not needed anymore

File: pulsar-flink-connector/src/main/java/org/apache/flink/streaming/connectors/pulsar/formats/protobufnative/PulsarProtobufNativeFormatFactory.java [hunk 2]
```diff
@@ -105,17 +105,34 @@ public ChangelogMode getChangelogMode() {
...
-
         validateTopic(context.getCatalogTable().getOptions());
-        String topic = context.getCatalogTable().getOptions().get(TOPIC.key());
-        // Maybe catalog mode
-        if (StringUtils.isNullOrWhitespaceOnly(topic)) {
-            final ObjectIdentifier table = context.getObjectIdentifier();
-            topic = TopicName.get(table.getDatabaseName() + "/" + table.getObjectName()).toString();
+
+        String topic;
+        if (isGenericTable(context)) {
```
Comment (imaffe) [inline pulsar-flink-connector/src/main/java/org/apache/flink/streaming/connectors/pulsar/formats/protobufnative/PulsarProtobufNativeFormatFactory.java:111]: There is a isGenericDatabase() method in PulsarCatalogSupport, which uses a different checking method

File: pulsar-flink-connector/src/main/java/org/apache/flink/streaming/connectors/pulsar/formats/protobufnative/PulsarProtobufNativeFormatFactory.java [hunk 3]
```diff
@@ -105,17 +105,34 @@ public ChangelogMode getChangelogMode() {
...
+     * Check if the table is an generic table. A generic table is defined as created using "CREATE
+     * TABLE". If the table is generic, when getting the table from catalog there will be an GENERIC
+     * option set to True
+     *
+     * @param context
+     * @return true if the topic is created by "CREATE TABLE" , false if the table is directly
+     *     mapped from pulsar topic
+     */
+    private boolean isGenericTable(DynamicTableFactory.Context context) {
+        final String isGeneric = context.getCatalogTable().getOptions().get(GENERIC.key());
```
Comment (imaffe) [inline pulsar-flink-connector/src/main/java/org/apache/flink/streaming/connectors/pulsar/formats/protobufnative/PulsarProtobufNativeFormatFactory.java:131]: Haven't tested if the GENERIC option will be set.  The GENERIC option is set when call getTable(), but will it be set when creating the table ?

Comment (nlu90) [reply pulsar-flink-connector/src/main/java/org/apache/flink/streaming/connectors/pulsar/formats/protobufnative/PulsarProtobufNativeFormatFactory.java:131]: The `GENERIC` is set whenever getTable() is called.
And any interaction with the table must call `getTable()` first.

File: pulsar-flink-connector/src/main/java/org/apache/flink/streaming/connectors/pulsar/formats/protobufnative/PulsarProtobufNativeFormatFactory.java [hunk 4]
```diff
@@ -105,17 +105,34 @@ public ChangelogMode getChangelogMode() {
...
+        return !StringUtils.isNullOrWhitespaceOnly(isGeneric)
+                && Boolean.TRUE.toString().equals(isGeneric);
```
Comment (nlu90) [inline pulsar-flink-connector/src/main/java/org/apache/flink/streaming/connectors/pulsar/formats/protobufnative/PulsarProtobufNativeFormatFactory.java:133]: `Boolean.parseBoolean(isGenrric)`

File: pulsar-flink-connector/src/main/java/org/apache/flink/streaming/connectors/pulsar/formats/protobufnative/PulsarProtobufNativeFormatFactory.java (hunk 2 above)
Comment (nlu90) [reply pulsar-flink-connector/src/main/java/org/apache/flink/streaming/connectors/pulsar/formats/protobufnative/PulsarProtobufNativeFormatFactory.java:111]: That's a catalog internal implementation. 

Other code should not assume any of its implementation. Just use the `isGeneric` table options, which is exposed as an indicator.

File: pulsar-flink-connector/src/main/java/org/apache/flink/streaming/connectors/pulsar/formats/protobufnative/PulsarProtobufNativeFormatFactory.java [hunk 5]
```diff
@@ -105,17 +105,34 @@ public ChangelogMode getChangelogMode() {
...
-            final ObjectIdentifier table = context.getObjectIdentifier();
-            topic = TopicName.get(table.getDatabaseName() + "/" + table.getObjectName()).toString();
+
+        String topic;
+        if (isGenericTable(context)) {
+            topic = context.getCatalogTable().getOptions().get(TOPIC.key());
+        } else {
+            String database = context.getObjectIdentifier().getDatabaseName();
+            String objectName = context.getObjectIdentifier().getObjectName();
+            topic = TopicName.get(database + "/" + objectName).toString();
```
Comment (nlu90) [inline pulsar-flink-connector/src/main/java/org/apache/flink/streaming/connectors/pulsar/formats/protobufnative/PulsarProtobufNativeFormatFactory.java:116]: you can reference https://github.com/streamnative/pulsar-flink/blob/master/pulsar-flink-connector/src/main/java/org/apache/flink/streaming/connectors/pulsar/catalog/util/PulsarCatalogSupport.java#L284 to generate a complete topic name

**Review by nlu90 — CHANGES_REQUESTED** [review]

File: pulsar-flink-connector/src/main/java/org/apache/flink/streaming/connectors/pulsar/formats/protobufnative/PulsarProtobufNativeFormatFactory.java [hunk 6]
```diff
@@ -105,17 +107,35 @@ public ChangelogMode getChangelogMode() {
...
         validateTopic(context.getCatalogTable().getOptions());
-        String topic = context.getCatalogTable().getOptions().get(TOPIC.key());
-        // Maybe catalog mode
-        if (StringUtils.isNullOrWhitespaceOnly(topic)) {
-            final ObjectIdentifier table = context.getObjectIdentifier();
-            topic = TopicName.get(table.getDatabaseName() + "/" + table.getObjectName()).toString();
+
+        String topic;
+        if (isGenericTable(context)) {
+            topic = context.getCatalogTable().getOptions().get(TOPIC.key());
```
Comment (imaffe) [inline pulsar-flink-connector/src/main/java/org/apache/flink/streaming/connectors/pulsar/formats/protobufnative/PulsarProtobufNativeFormatFactory.java:114]: If it is a generic table

1. use getCatalogTable() refers to the orignal native pulsar topic (not in __flink_catalog tenant)
2. objectNameToTopicName() refers to the mapped topic (under __flink_catalog tenant)
I think here what we want is the topic carrying flink schema ? So maybe we should use 2nd approach ?

File: pulsar-flink-connector/src/main/java/org/apache/flink/streaming/connectors/pulsar/catalog/util/PulsarCatalogSupport.java [hunk 7]
```diff
@@ -283,7 +283,6 @@ private String objectNameToTopicName(ObjectPath objectPath) {
 
         NamespaceName ns = NamespaceName.get(database);
         TopicName fullName = TopicName.get(TopicDomain.persistent.toString(), ns, topic);
-
```
Comment (imaffe) [inline pulsar-flink-connector/src/main/java/org/apache/flink/streaming/connectors/pulsar/catalog/util/PulsarCatalogSupport.java:286]: This deletion is unnecessary.

**Comment by imaffe** [comment]
Turns out the shaded jar didn't not include protobuf jars. (which i think should be included)
Adding this line solve the ClassNotFoundError

**Review by nlu90 — APPROVED** [review]

**Pull request merged**